src/task/future.rs

```diff
@@ -0,0 +1,388 @@
+    remote: Remote,
+    extras: UnsafeCell<Extras>,
+}
+
+/// A [`Future`] task cell.
+pub struct TaskCell<Remote, Extras>(Arc<Task<Remote, Extras>>);
+
+// Safety: It is ensured that `future` and `extras` are always accessed by
+// only one thread at the same time.
+unsafe impl<Remote: Sync, Extras> Sync for Task<Remote, Extras> {}
+
+impl<Remote, Extras> fmt::Debug for TaskCell<Remote, Extras> {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        "future::TaskCell".fmt(f)
+    }
+}
+
+const WAITING: u8 = 0; // --> POLLING
+const POLLING: u8 = 1; // --> WAITING, REPOLL, or COMPLETE
+const REPOLL: u8 = 2; // --> POLLING
+const COMPLETE: u8 = 3; // No transitions out
+
+impl<Remote, Extras> TaskCell<Remote, Extras> {
+    /// Creates a [`Future`] task cell.
+    pub fn new<F: Future<Output = ()> + Send + 'static>(
+        future: F,
+        remote: Remote,
+        extras: Extras,
+    ) -> Self {
+        TaskCell(Arc::new(Task {
+            status: AtomicU8::new(WAITING),
+            future: UnsafeCell::new(Box::pin(future)),
+            remote,
+            extras: UnsafeCell::new(extras),
+        }))
+    }
+}
+
+impl<Remote, Extras> crate::queue::TaskCell for TaskCell<Remote, Extras> {
+    type Extras = Extras;
+
+    fn mut_extras(&mut self) -> &mut Self::Extras {
+        unsafe { &mut *self.0.extras.get() }
+    }
+}
+
+#[inline]
+unsafe fn waker<Remote, Extras>(task: *const Task<Remote, Extras>) -> Waker
+where
+    Remote: RemoteSpawn<TaskCell = TaskCell<Remote, Extras>>,
+{
+    Waker::from_raw(RawWaker::new(
+        task as *const (),
+        &RawWakerVTable::new(
+            clone_raw::<Remote, Extras>,
+            wake_raw::<Remote, Extras>,
+            wake_ref_raw::<Remote, Extras>,
+            drop_raw::<Remote, Extras>,
+        ),
+    ))
+}
+
+#[inline]
+unsafe fn clone_raw<Remote, Extras>(this: *const ()) -> RawWaker
+where
+    Remote: RemoteSpawn<TaskCell = TaskCell<Remote, Extras>>,
+{
+    let task_cell = clone_task(this as *const Task<Remote, Extras>);
+    RawWaker::new(
+        Arc::into_raw(task_cell.0) as *const (),
+        &RawWakerVTable::new(
+            clone_raw::<Remote, Extras>,
+            wake_raw::<Remote, Extras>,
+            wake_ref_raw::<Remote, Extras>,
+            drop_raw::<Remote, Extras>,
+        ),
+    )
+}
+
+#[inline]
+unsafe fn drop_raw<Remote, Extras>(this: *const ())
+where
+    Remote: RemoteSpawn<TaskCell = TaskCell<Remote, Extras>>,
+{
+    drop(task_cell(this as *const Task<Remote, Extras>))
+}
+
+unsafe fn wake_impl<Remote, Extras>(task_cell: &TaskCell<Remote, Extras>)
+where
+    Remote: RemoteSpawn<TaskCell = TaskCell<Remote, Extras>>,
+{
+    let task = &task_cell.0;
+    let mut status = task.status.load(SeqCst);
+    loop {
+        match status {
+            WAITING => {
+                match task
+                    .status
+                    .compare_exchange(WAITING, POLLING, SeqCst, SeqCst)
```

Comment:
I find it quite confusing that marking woken up task as `POLLING`, because it actually has not polled it yet.

I suggest to use following state instead. 
```
const NOTIFIED: u8 = 1;
const IDLE: u8 = 2;
const POLLING: u8 = 3;
const COMPLETED: u8 = 4;
```

Ref: https://github.com/tikv/grpc-rs/blob/0c16249ca5afabbbf308e9d6dda9804ffcf0535a/src/task/executor.rs#L73-L84

Reply:
Updated. PTAL again.